File: java/src/main/java/ev/projects/PatternLine.java
package ev.projects;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class PatternLine {
    private final int size;
    private int tileCount;
    private Tile tile;

    public PatternLine(int size) {
        this.size = size;
        tileCount = 0;
        tile = null;
    }

    public int add(Tile tile, int count) {
        if(tileCount > 0 && this.tile != tile) {
            throw new ActionNotAllowedException(
                    "Tile(s) with " + this.tile + " colour is on the pattern line. Can't add a tile with " + tile + " colour."
            );
        }
        this.tile = tile;
        int overfill = Math.max(0, tileCount + count - size);
        tileCount = Math.min(size, tileCount + count);
        return overfill;
    }

    public boolean isFilled() {
        return tileCount == size;
    }

    public int tileCount() {
        return tileCount;
    }

    public Tile tile() {
        return tile;
    }

    public void clear() {
        tileCount = 0;
        tile = null;
    }

    @Override
    public String toString() {
        return Optional.ofNullable(tile).map(t -> (t + " ").repeat(tileCount)).orElse("")
                + "E ".repeat(size - tileCount);
    }

    public List<String> jsonList() {
        return Stream.generate(() -> tile.toString())
                    .limit(tileCount)
                    .collect(Collectors.toList());
    }
}

File: python/src/pattern_line.py
from src.action_not_allowed_exception import ActionNotAllowedException

class PatternLine:
    def __init__(self, size):
        self.size = size
        self.tile_count = 0
        self.tile = None

    def add(self, tile, count):
        if self.tile_count > 0 and self.tile != tile:
            raise ActionNotAllowedException(
                f"Tile(s) with {self.tile} colour is on the pattern line. Can't add a tile with {tile} colour."
            )
        self.tile = tile
        overfill = max(0, self.tile_count + count - self.size)
        self.tile_count = min(self.size, self.tile_count + count)
        return overfill

    def is_filled(self):
        return self.tile_count == self.size

    def clear(self):
        self.tile_count = 0
        self.tile = None

    def __str__(self):
        return (
            (f"{self.tile} " * self.tile_count if self.tile else "")
            + "E " * (self.size - self.tile_count)
        )

    def json_list(self):
        return [str(self.tile)] * self.tile_count if self.tile else []